### PR TITLE
Move all plugins and their dependencies to the front

### PR DESCRIPTION
**Pull request by balbuf**
Currently plugins are only moved to the front to be installed first when they don't have dependencies beyond platform requirements and `composer-plugin-api`. This causes an issue for our plugin [`oomphinc/composer-installers-extender`](https://github.com/oomphinc/composer-installers-extender) which is built atop `composer/composer-installers` and thus requires it. Users are seeing the same issue as https://github.com/composer/composer/issues/1147, see: https://github.com/oomphinc/composer-installers-extender/issues/6.

What this does is instead of skipping plugins that have additional dependencies, it moves the plugin and those dependencies to the front as well. This works on the assumption that install/update operations for a package's dependencies come before the package that required it. I'm not sure if that is a completely safe assumption to make, but it seemed to be the case in my testing.

I'm not sure if there were other reasons why plugins that had additional dependencies were skipped over from being moved in the first place; are there any other implications here that I am neglecting? As well, do you think this approach guaranteed to move plugins and all their dependencies properly or do you think more robust logic is necessary?

Please let me know your thoughts!

### CODE REVIEW

**Comment by phenaproxima** [comment]
Please, +1 for this.

**Comment by balbuf** [comment]
In terms of my wondering whether a package that depends on another is consistently installed after its dependency, it seems that is entirely the point of how the installation order is determined. At least based on what's been said in the past by some reputable sources:

> The only constraint used when constructing a transaction is that 
packages which are required by another package are installed before the 
package that requires them. 

https://groups.google.com/d/msg/composer-dev/a14g7NiELv0/RFCDYvjJlLcJ

> the rule to sort the packages for installations is that dependencies of a package are always installed before the package itself (except if you have a circular dependency in your packages of course as this would be impossible)

https://github.com/composer/composer/issues/2972#issuecomment-42497953

So if that is indeed true, then when starting from the end of the transaction list and working backwards, you should always first encounter packages that are not depended on by anything but the root package itself, followed by any of their direct dependencies, which in turn are always followed by their direct dependencies and so on. In other words, the algorithm this PR uses should be sound, barring any anomalies I am neglecting.

File: src/Composer/Installer.php [hunk 1]
```diff
@@ -754,19 +755,15 @@ private function movePluginsToFront(array $operations)
...
-                if (!count($requires)) {
-                    $installerOps[] = $op;
-                    unset($operations[$idx]);
-                }
+            if ($package->getType() === 'composer-plugin' 
+                || $package->getType() === 'composer-installer'
+                || in_array($package->getName(), $installerRequires)
+            ) {
+                // capture the requirements for this package so those packages will be moved up as well
+                $installerRequires = array_merge($installerRequires, array_keys($package->getRequires()));
```
Comment (naderman) [inline src/Composer/Installer.php]: this needs to take provider names into account, not just requires

File: tests/Composer/Test/Fixtures/installer/plugins-are-installed-first.test [hunk 2]
```diff
@@ -25,7 +25,7 @@ Composer installers are installed first if they have no meaningful requirements
 install
 --EXPECT--
 Installing inst (1.0.0)
-Installing inst-with-req (1.0.0)
-Installing pkg (1.0.0)
 Installing pkg2 (1.0.0)
 Installing inst-with-req2 (1.0.0)
+Installing inst-with-req (1.0.0)
```
Comment (naderman) [inline tests/Composer/Test/Fixtures/installer/plugins-are-installed-first.test]: this plugin is now not loaded for pkg2 installation anymore, so we should try and still move any plugin without dependencies to the very front to allow them to be loaded for every non-plugin dependency

Comment (balbuf) [inline tests/Composer/Test/Fixtures/installer/plugins-are-installed-first.test]: @naderman just to confirm, what you are saying is to move all plugins with no dependencies to the very front, followed by any plugins with their dependencies, followed by all other dependencies in their previously determined order (i.e. left where they are)?

File: src/Composer/Installer.php (hunk 1 above)
Comment (balbuf) [inline src/Composer/Installer.php]: @naderman does that mean we need to consider "replaces" as well?

Comment (balbuf) [inline src/Composer/Installer.php]: @naderman is this what you were getting at? see: https://github.com/composer/composer/pull/6371/commits/4cda7e0a44d608f5bb6071003790046594535549

**Comment by balbuf** [comment]
@naderman I've made changes based on what I understood your requests to be - let me know what you think!

File: src/Composer/Installer.php (hunk 1 above)
Comment (naderman) [inline src/Composer/Installer.php]: @balbuf yeah that's what I had in mind.

File: tests/Composer/Test/Fixtures/installer/plugins-are-installed-first.test (hunk 2 above)
Comment (naderman) [inline tests/Composer/Test/Fixtures/installer/plugins-are-installed-first.test]: Exactly :-)

**Comment by balbuf** [comment]
@naderman @Seldaek thanks for merging this in! So that I can provide an update on https://github.com/oomphinc/composer-installers-extender/issues/6, when do you think this will makes its way into a release?

**Comment by Seldaek** [comment]
@balbuf took a while sorry but 1.5 is out with this change

**Comment by balbuf** [comment]
Thanks, @Seldaek !

**Pull request merged**